### PR TITLE
feat(@formatjs/intl-segmenter): Intl.Segmenter polyfill (stage 4)

### PR DESCRIPTION
**Pull request by matijagaspar**
This PR aims to implement: [proposal-intl-segmenter](https://github.com/tc39/proposal-intl-segmenter) 

## Basic Information

[UAX29](https://unicode.org/reports/tr41/tr41-30.html#UAX29) defines the general rules for grapheme, sentence and word splitting. Those rules (together with locale specific overrides) are defined in "machine readable" code in [CLDR](https://github.com/unicode-org/cldr/tree/main/common/segments). Following the UAX29 recommendation of making use of the CLDR, this implementation attempts to convert those rules (they are in unicode regex with added variables and break notation) into es5 compatible regex.

## Implemementation details

[scripts/generate-cldr-segmentation-rules](https://github.com/formatjs/formatjs/blob/d82e526ec666dd41855e7e43dcd1efad3212f066/packages/intl-segmenter/scripts/generate-cldr-segmentation-rules.ts#L304) generates a es5 regex compatible ruleset from `cldr-segments-full` files. It uses [UCD](https://unicode.org/ucd/) [files](https://unicode.org/Public/UCD/latest/ucd/) to parse the needed codepoints for the rules. It heavily relies on [regexpu-core](https://github.com/mathiasbynens/regexpu-core) for the regex transformations and some custom transformations to address other incompatibilities that `regexpu-core` does not handle.

[The segmenter iterator](https://github.com/formatjs/formatjs/blob/d82e526ec666dd41855e7e43dcd1efad3212f066/packages/intl-segmenter/src/segmenter.ts) than loops through each character of a string and checks each CLDR rule (executed in numerical order) if before regex and after regex match. If a rule matches return the break (true or false depending on the rule).

Additionally there are hardcoded rules (defined in uax29 but not part of CLDR):

- Returns true at index 0 `[0.2] SOT ÷`
- Returns true at index==input.length `[0.3] EOT ÷`
- Returns true if no other rule matches `[999] Any ÷ Any`

And this implementation specific artificial rules:

- ES5 regex is not unicode aware, return false if previous character is part of a surrogate pair `[0.1]`
- artificial rule `[0.4]` for suppressions defined in CLDR (for specific locale overrides)

## Spec compliance
The implementation passes all of the latest [UCD tests](https://unicode.org/reports/tr41/tr41-26.html#Tests29) and almost all test262 (see:  [**Failing test262 tests**](#test262))

## Potential alternative solutionos

- ~~using Java implementation of unicode CLDR rule transformation to java compatible~~ not feasible as there would still be need to transform the java regex to es5 compatible.
- Hardcode the rules as javascript logic (still making use of UCD files to gather all the relevant codepoints). Not ideal, due to higher maintenance if any update to cldrs.
- Instead of regex transformation, implement a custom interpreter of the CLDR rules. Might be a better approach, however would be more complex, as it would essentially require to create a new custom regex like parser in javascript.

## Various Notes and information regarding unicode segmentation

Throughout the work on this I have made various notes collected in: [NOTES.md](https://github.com/formatjs/formatjs/blob/d82e526ec666dd41855e7e43dcd1efad3212f066/packages/intl-segmenter/NOTES.md)

They are mostly for my own reference, but I left them in for now because I believe it's a good source of information regarding this topic

### Failing test262 tests<a id='test262'></a>

- Supported locales related tests (currently I added supported locales found in CLDR but tests expect other locales to be avilible):

  - [constructor/supportedLocalesOf/locales-specific.js](https://github.com/tc39/test262/blob/main/test/intl402/Segmenter/constructor/supportedLocalesOf/locales-specific.js) `sr` locale not in segmenter availible locales, not sure about the correct approach.
  - [constructor/constructor/locales-valid.js](https://github.com/tc39/test262/blob/main/test/intl402/Segmenter/constructor/constructor/locales-valid.js)
  - [prototype/resolvedOptions/type-without-lbs.js](https://github.com/tc39/test262/blob/main/test/intl402/Segmenter/prototype/resolvedOptions/type-without-lbs.js)
  - [constructor/supportedLocalesOf/basic.js](https://github.com/tc39/test262/blob/main/test/intl402/Segmenter/constructor/supportedLocalesOf/basic.js)

- [constructor/constructor/options-undefined.js](https://github.com/tc39/test262/blob/main/test/intl402/Segmenter/constructor/constructor/options-undefined.js) set to excluded, similarly to other packages

### CLDR Transform issues

- Grapheme Rule 13 and Word Rule 16 use negated character class for Regional indicator (which is a surrogate pair), currently the code does not handle this transformation dynamically and there is a [hardcoded replacements](https://github.com/formatjs/formatjs/blob/5944a658ca392fba9d0f4b256a8cb35b0197eb13/packages/intl-segmenter/scripts/generate-cldr-segmentation-rules.ts#L26) for it.
- Variable `$STerm` in `SentenceBreak` for `el`  locale seems to have an invalid definition, there is a hardcoded variable [replacement with the fix](https://github.com/formatjs/formatjs/blob/5944a658ca392fba9d0f4b256a8cb35b0197eb13/packages/intl-segmenter/scripts/generate-cldr-segmentation-rules.ts#L35).

## Benchmark

Ran on Node v18.12.1

### Built in v8 Intl.Segmenter

```
Locale: en
Input string size: 8 KB
segment_cached_grapheme_collect_all x 22.42 ops/sec ±9.16% (40 runs sampled)
segment_cached_word_collect_all x 64.16 ops/sec ±7.71% (57 runs sampled)
segment_cached_sentence_collect_all x 2,899 ops/sec ±1.35% (90 runs sampled)
segment_cached_grapheme_containing x 186,612 ops/sec ±2.20% (88 runs sampled)
segment_cached_word_containing x 174,066 ops/sec ±1.18% (91 runs sampled)
segment_cached_sentence_containing x 128,220 ops/sec ±0.70% (94 runs sampled)
new_segmenter_grapheme x 207,883 ops/sec ±8.65% (79 runs sampled)
new_segmenter_word x 163,141 ops/sec ±17.94% (78 runs sampled)
new_segmenter_sentence x 213,711 ops/sec ±8.45% (80 runs sampled)
```

### Polyfilled segmenter

```
Polyfill on Node 18:
Locale: en
Input string size: 8 KB
segment_cached_grapheme_collect_all x 331 ops/sec ±0.41% (92 runs sampled)
segment_cached_word_collect_all x 256 ops/sec ±1.63% (89 runs sampled)
segment_cached_sentence_collect_all x 45.88 ops/sec ±1.36% (61 runs sampled)
segment_cached_grapheme_containing x 1,320,075 ops/sec ±0.65% (97 runs sampled)
segment_cached_word_containing x 5,082 ops/sec ±0.77% (83 runs sampled)
segment_cached_sentence_containing x 934 ops/sec ±2.47% (74 runs sampled)
new_segmenter_grapheme x 6,384 ops/sec ±1.07% (93 runs sampled)
new_segmenter_word x 1,404 ops/sec ±1.72% (93 runs sampled)
new_segmenter_sentence x 431 ops/sec ±18.60% (21 runs sampled)
```

When using the iterator the polyfilled version performs better than the built in Intl.Segmenter except on sentence. Larger the text, larger the difference.
Random access through `containing()` is significantly slower compared to the built in v8 `Intl.Segmenter` except for grapheme due. This is because the need to looping backward and forward to find the boundaries.

Instantiation on the polyfill is slow, due to the need of preparing the rule set for selected locale and granularity.

cc @micheleriva

### CODE REVIEW

File: packages/intl-segmenter/scripts/generate-cldr-segmentation-rules.ts [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+    if (
+      'segmentRules' in segmentationTypeValue &&
+      segmentationTypeValue.segmentRules
+    ) {
+      for (const [ruleNr, ruleString] of Object.entries(
+        segmentationTypeValue.segmentRules
+      )) {
+        //replace negated character class to negative lookahead inside noncapturing group (doesn't work correctly, needs fixing)
+        //somewhat hacky way to transform the rule regex, can't use regexpu because we need to preserve the varialbes
+        //TODO: needs fixing
```
Comment (micheleriva) [inline packages/intl-segmenter/scripts/generate-cldr-segmentation-rules.ts]: Is there a way to fix this before moving the PR into a review stage?

Comment (matijagaspar) [reply packages/intl-segmenter/scripts/generate-cldr-segmentation-rules.ts]: I have updated the comment to make it more clear.
The negated character classes in the rules are currently handled as described in PR description under "CLDR Transform issues"

**Comment by longlho** [comment]
is this wip?

**Comment by matijagaspar** [comment]
Its ready for review @longlho 

If is regarding the failing 262 tests, I need some input regarding that. Tests are checking for available locales that are not in CLDR for segmentation, so should they still be in the list of "availible locales" or something else?

I can quickly change based on the answer.

Regarding **CLDR Transform issues**, there are workarounds in place which at this stage I believe are ok. To dynamically handle rule 13 and 16 transformation, it would add considerable complexity that I do not think is worth it. The 2nd bullet point regarding `el` locale, I believe is a "bug". I have opened the issue on [CLDR Jira](https://unicode-org.atlassian.net/browse/CLDR-16176).

**Comment by longlho** [comment]
kk thanks @matijagaspar given that this PR is huge I'll merge as is and we can follow up on fixes later.

**Comment by matijagaspar** [comment]
@longlho what should be the next steps? (regarding fixes, and how can I assist)

**Pull request merged**